File: BST/1_search_in_BST.cpp
#include<iostream>
#include<queue>
#include<bits/stdc++.h>
using namespace std;
// Constructing Binary Tree
// create node of tree 
struct Node{

    int key;
    Node *left;
    Node *right;
    Node(int k){
        key=k;
        left=NULL;
        right=NULL;
    }
};
// Search in Binary search tree: recursive approch---> time=O(log N)
bool search(Node *root, int x)
{
    // base case
    if(root==NULL)
        return false;
    else if(root->key==x)
        return true;
    else if(root->key>x)
        return search(root->left,x);
    else
        return search(root->right,x);
}
bool searchITR(Node *root, int x)
{
    // base case
    if(root==NULL)
        return false;
    while(root!=NULL)
    {
        if(root->key==x)
            return true;
        else if(root->key>x)
            return search(root->left,x);
        else
            return search(root->right,x);
    }
    return false;
}

int main(){
    // Node *root=NULL;    // empty tree
    Node *root=new Node(15);
    root->left=new Node(5);
    root->right=new Node(20);
    root->left->left=new Node(3);
    root->right->right=new Node(80);
    root->right->left=new Node(18);
    root->right->left->left=new Node(16);
    cout<<search(root,2)<<endl;
    cout<<searchITR(root,2)<<endl;
    return 0;
}

File: code_java/self pased DSA/BST/1_search_in_BST.java
/**
 * Problem: Search for a given key in a Binary Search Tree (BST).
 *
 * Definition:
 * A Binary Search Tree (BST) is a binary tree in which each node has the following properties:
 * - The left subtree of a node contains only nodes with keys less than the node’s key.
 * - The right subtree of a node contains only nodes with keys greater than the node’s key.
 *
 * Goal:
 * Given the root of a BST and a key, determine whether the key exists in the tree.
 *
 * Approach:
 * We use a recursive method to perform the search:
 * - If the current node is null, the key is not present.
 * - If the key equals the current node’s key, return true.
 * - If the key is less than the current node’s key, search the left subtree.
 * - Otherwise, search the right subtree.
 *
 * Why this approach:
 * This approach leverages the BST property to eliminate half of the tree at each step,
 * reducing the time complexity to O(h), where h is the height of the tree.
 * In a balanced BST, this gives O(log n) performance.
 *
 * This method is efficient because it narrows the search space using the inherent ordering of the BST.
 */

class Node {
    int key;
    Node left, right;

    Node(int key) {
        this.key = key;
        left = right = null;
    }
}

public class BSTSearch {

    // Recursive method to search in BST
    public static boolean search(Node root, int x) {
        if (root == null) return false;
        if (root.key == x) return true;
        else if (x < root.key) return search(root.left, x);
        else return search(root.right, x);
    }

    public static void main(String[] args) {
        Node root = new Node(15);
        root.left = new Node(10);
        root.right = new Node(20);
        root.left.left = new Node(8);
        root.left.right = new Node(12);

        int target = 10;
        System.out.println("Found " + target + "? " + search(root, target));
    }
}